### PR TITLE
Introduce a <model> loading process

### PR DESCRIPTION
**Pull request by hortont424**
#### fec04ae915555611a91c8612dd87160356a41c8f
<pre>
Introduce a &lt;model&gt; loading process
<a href="https://bugs.webkit.org/show_bug.cgi?id=243874">https://bugs.webkit.org/show_bug.cgi?id=243874</a>

Reviewed by NOBODY (OOPS!).

In order to allow parsing of complex model formats with a custom (tight) sandbox,
introduce a separate process for loading/parsing &lt;model&gt;s. (For now the sandbox
is just cloned from the Web Content process, but we can reduce the scope later).

For now, a single model process is scoped to a single Web content process, and
remains running once launched. Future patches will more closely tie its
lifetime to live &lt;model&gt; elements.

The overall process structure is cloned from the GPU process and reduced to
the bare minimum of things that we will likely need (e.g. I left the visibility
propagation code, process suspension and low memory handlers, etc.)

* .gitignore:
Remove a confusingly widely scoped gitignore. It doesn&apos;t seem to be a file extension
that Xcode emits anymore, and conflicts with the sandbox profiles below.

* Source/WTF/Scripts/Preferences/WebPreferencesInternal.yaml:
Add a off-by-default preference for using the Model-process backed &lt;model&gt; element.

* Source/WTF/wtf/PlatformEnableCocoa.h:
Add an on-by-default-on-Cocoa-platforms compile-time flag for the Model process.

* Source/WebCore/platform/RuntimeApplicationChecks.cpp:
(WebCore::processTypeDescription):
* Source/WebCore/platform/RuntimeApplicationChecks.h:
* Source/WebKit/Configurations/ModelService.xcconfig: Added.
* Source/WebKit/Configurations/WebKit.xcconfig:
* Source/WebKit/DerivedSources-input.xcfilelist:
* Source/WebKit/DerivedSources-output.xcfilelist:
* Source/WebKit/DerivedSources.make:
* Source/WebKit/GPUProcess/GPUProcess.cpp:
* Source/WebKit/ModelProcess/EntryPoint/Cocoa/XPCService/ModelService/Info-OSX.plist: Added.
* Source/WebKit/ModelProcess/EntryPoint/Cocoa/XPCService/ModelService/Info-iOS.plist: Added.
* Source/WebKit/ModelProcess/EntryPoint/Cocoa/XPCService/ModelServiceEntryPoint.mm: Copied from Source/WebKit/WebProcess/Model/WebModelPlayerProvider.cpp.
(WebKit::ModelServiceInitializerDelegate::ModelServiceInitializerDelegate):
(WebKit::initializeAuxiliaryProcess&lt;ModelProcess&gt;):
(MODEL_SERVICE_INITIALIZER):
* Source/WebKit/ModelProcess/ModelConnectionToWebProcess.cpp: Added.
(WebKit::ModelConnectionToWebProcess::create):
(WebKit::ModelConnectionToWebProcess::ModelConnectionToWebProcess):
(WebKit::ModelConnectionToWebProcess::~ModelConnectionToWebProcess):
(WebKit::ModelConnectionToWebProcess::didClose):
(WebKit::ModelConnectionToWebProcess::createVisibilityPropagationContextForPage):
(WebKit::ModelConnectionToWebProcess::destroyVisibilityPropagationContextForPage):
(WebKit::ModelConnectionToWebProcess::configureLoggingChannel):
(WebKit::ModelConnectionToWebProcess::allowsExitUnderMemoryPressure const):
(WebKit::ModelConnectionToWebProcess::logger):
(WebKit::ModelConnectionToWebProcess::didReceiveInvalidMessage):
(WebKit::ModelConnectionToWebProcess::lowMemoryHandler):
(WebKit::ModelConnectionToWebProcess::dispatchMessage):
(WebKit::ModelConnectionToWebProcess::dispatchSyncMessage):
* Source/WebKit/ModelProcess/ModelConnectionToWebProcess.h: Added.
(WebKit::ModelConnectionToWebProcess::connection):
(WebKit::ModelConnectionToWebProcess::messageReceiverMap):
(WebKit::ModelConnectionToWebProcess::modelProcess):
(WebKit::ModelConnectionToWebProcess::webProcessIdentifier const):
(WebKit::ModelConnectionToWebProcess::sessionID const):
(WebKit::ModelConnectionToWebProcess::webProcessIdentity const):
(WebKit::ModelConnectionToWebProcess::objectCountForTesting):
* Source/WebKit/ModelProcess/ModelConnectionToWebProcess.messages.in: Added.
* Source/WebKit/ModelProcess/ModelProcess.cpp: Added.
(WebKit::ModelProcess::ModelProcess):
(WebKit::ModelProcess::~ModelProcess):
(WebKit::ModelProcess::didReceiveMessage):
(WebKit::asConnectionIdentifier):
(WebKit::ModelProcess::createModelConnectionToWebProcess):
(WebKit::ModelProcess::removeModelConnectionToWebProcess):
(WebKit::ModelProcess::connectionToWebProcessClosed):
(WebKit::ModelProcess::shouldTerminate):
(WebKit::ModelProcess::canExitUnderMemoryPressure const):
(WebKit::ModelProcess::tryExitIfUnusedAndUnderMemoryPressure):
(WebKit::ModelProcess::tryExitIfUnused):
(WebKit::ModelProcess::lowMemoryHandler):
(WebKit::ModelProcess::initializeModelProcess):
(WebKit::ModelProcess::prepareToSuspend):
(WebKit::ModelProcess::processDidResume):
(WebKit::ModelProcess::resume):
(WebKit::ModelProcess::webProcessConnection const):
(WebKit::ModelProcess::webProcessConnectionCountForTesting):
(WebKit::ModelProcess::addSession):
(WebKit::ModelProcess::removeSession):
* Source/WebKit/ModelProcess/ModelProcess.h: Added.
(WebKit::ModelProcess::applicationVisibleName const):
* Source/WebKit/ModelProcess/ModelProcess.messages.in: Added.
* Source/WebKit/ModelProcess/ModelProcessCreationParameters.cpp: Added.
(WebKit::ModelProcessCreationParameters::encode const):
(WebKit::ModelProcessCreationParameters::decode):
* Source/WebKit/ModelProcess/ModelProcessCreationParameters.h: Added.
* Source/WebKit/ModelProcess/cocoa/ModelProcessCocoa.mm: Added.
(WebKit::ModelProcess::notifyPreferencesChanged):
(WebKit::ModelProcess::dispatchSimulatedNotificationsForPreferenceChange):
* Source/WebKit/ModelProcess/ios/ModelProcessIOS.mm: Added.
(WebKit::ModelProcess::initializeProcess):
(WebKit::ModelProcess::initializeProcessName):
(WebKit::ModelProcess::initializeSandbox):
* Source/WebKit/ModelProcess/mac/ModelProcessMac.mm: Added.
(WebKit::ModelProcess::initializeProcess):
(WebKit::ModelProcess::initializeProcessName):
(WebKit::ModelProcess::initializeSandbox):
* Source/WebKit/ModelProcess/mac/com.apple.WebKit.ModelProcess.sb.in: Added.
* Source/WebKit/Platform/ExtraPrivateSymbolsForTAPI.h:
* Source/WebKit/Resources/SandboxProfiles/ios/com.apple.WebKit.Model.sb.in: Added.
* Source/WebKit/Scripts/process-entitlements.sh:
* Source/WebKit/Shared/Cocoa/DefaultWebBrowserChecks.mm:
(WebKit::isInWebKitChildProcess):
* Source/WebKit/Shared/EntryPointUtilities/Cocoa/XPCService/XPCServiceEntryPoint.h:
* Source/WebKit/Shared/EntryPointUtilities/Cocoa/XPCService/XPCServiceMain.mm:
(WebKit::XPCServiceEventHandler):
* Source/WebKit/Shared/ModelProcessConnectionParameters.h: Added.
(WebKit::ModelProcessConnectionParameters::encode const):
(WebKit::ModelProcessConnectionParameters::decode):
* Source/WebKit/Shared/mac/AuxiliaryProcessMac.mm:
(WebKit::processStorageClass):
(WebKit::sandboxDirectory):
* Source/WebKit/Sources.txt:
* Source/WebKit/SourcesCocoa.txt:
* Source/WebKit/UIProcess/API/Cocoa/WKWebView.mm:
(-[WKWebView _modelProcessIdentifier]):
* Source/WebKit/UIProcess/API/Cocoa/WKWebViewPrivate.h:
* Source/WebKit/UIProcess/AuxiliaryProcessProxy.cpp:
(WebKit::AuxiliaryProcessProxy::getLaunchOptions):
* Source/WebKit/UIProcess/Cocoa/PageClientImplCocoa.h:
* Source/WebKit/UIProcess/Cocoa/PageClientImplCocoa.mm:
(WebKit::PageClientImplCocoa::modelProcessDidFinishLaunching):
(WebKit::PageClientImplCocoa::modelProcessDidExit):
* Source/WebKit/UIProcess/Cocoa/WebPageProxyCocoa.mm:
(WebKit::WebPageProxy::didCreateContextInModelProcessForVisibilityPropagation):
* Source/WebKit/UIProcess/Launcher/ProcessLauncher.h:
* Source/WebKit/UIProcess/Launcher/cocoa/ProcessLauncherCocoa.mm:
(WebKit::serviceName):
(WebKit::shouldLeakBoost):
* Source/WebKit/UIProcess/Model/ModelProcessProxy.cpp: Added.
(WebKit::ModelProcessProxy::create):
(WebKit::ModelProcessProxy::ModelProcessProxy):
(WebKit::ModelProcessProxy::getLaunchOptions):
(WebKit::ModelProcessProxy::connectionWillOpen):
(WebKit::ModelProcessProxy::processWillShutDown):
(WebKit::ModelProcessProxy::createModelProcessConnection):
(WebKit::ModelProcessProxy::modelProcessExited):
(WebKit::ModelProcessProxy::processIsReadyToExit):
(WebKit::ModelProcessProxy::addSession):
(WebKit::ModelProcessProxy::removeSession):
(WebKit::ModelProcessProxy::terminateForTesting):
(WebKit::ModelProcessProxy::webProcessConnectionCountForTesting):
(WebKit::ModelProcessProxy::didClose):
(WebKit::ModelProcessProxy::didReceiveInvalidMessage):
(WebKit::ModelProcessProxy::didFinishLaunching):
(WebKit::ModelProcessProxy::updateProcessAssertion):
(WebKit::ModelProcessProxy::sendPrepareToSuspend):
(WebKit::ModelProcessProxy::sendProcessDidResume):
(WebKit::ModelProcessProxy::didCreateContextForVisibilityPropagation):
(WebKit::ModelProcessProxy::didBecomeUnresponsive):
* Source/WebKit/UIProcess/Model/ModelProcessProxy.h: Added.
* Source/WebKit/UIProcess/Model/ModelProcessProxy.messages.in: Added.
* Source/WebKit/UIProcess/PageClient.h:
(WebKit::PageClient::didCreateContextInModelProcessForVisibilityPropagation):
(WebKit::PageClient::modelProcessDidFinishLaunching):
(WebKit::PageClient::modelProcessDidExit):
* Source/WebKit/UIProcess/ProvisionalPageProxy.cpp:
(WebKit::ProvisionalPageProxy::didCreateContextInModelProcessForVisibilityPropagation):
* Source/WebKit/UIProcess/ProvisionalPageProxy.h:
(WebKit::ProvisionalPageProxy::contextIDForVisibilityPropagationInModelProcess const):
* Source/WebKit/UIProcess/WebPageProxy.cpp:
(WebKit::WebPageProxy::modelProcessIdentifier const):
(WebKit::WebPageProxy::modelProcessDidFinishLaunching):
(WebKit::WebPageProxy::modelProcessExited):
* Source/WebKit/UIProcess/WebPageProxy.h:
* Source/WebKit/UIProcess/WebProcessPool.h:
* Source/WebKit/UIProcess/WebProcessProxy.cpp:
(WebKit::WebProcessProxy::createModelProcessConnection):
(WebKit::WebProcessProxy::modelProcessDidFinishLaunching):
(WebKit::WebProcessProxy::modelProcessExited):
(WebKit::WebProcessProxy::ensureModelProcess):
(WebKit::WebProcessProxy::didSetAssertionType):
(WebKit::WebProcessProxy::currentProcessAssertionType):
* Source/WebKit/UIProcess/WebProcessProxy.h:
(WebKit::WebProcessProxy::modelProcess const):
* Source/WebKit/UIProcess/WebProcessProxy.messages.in:
* Source/WebKit/UIProcess/ios/PageClientImplIOS.h:
* Source/WebKit/UIProcess/ios/PageClientImplIOS.mm:
(WebKit::PageClientImpl::didCreateContextInModelProcessForVisibilityPropagation):
(WebKit::PageClientImpl::modelProcessDidExit):
* Source/WebKit/UIProcess/ios/WKContentView.h:
* Source/WebKit/UIProcess/ios/WKContentView.mm:
(-[WKContentView _commonInitializationWithProcessPool:configuration:]):
(-[WKContentView _setupVisibilityPropagationViewForModelProcess]):
(-[WKContentView _removeVisibilityPropagationViewForModelProcess]):
(-[WKContentView _modelProcessDidExit]):
(-[WKContentView _didRelaunchProcess]):
(-[WKContentView _modelProcessDidCreateContextForVisibilityPropagation]):
* Source/WebKit/WebKit.xcodeproj/project.pbxproj:
* Source/WebKit/WebProcess/Model/ModelProcessConnection.cpp: Added.
(WebKit::getModelProcessConnectionParameters):
(WebKit::ModelProcessConnection::create):
(WebKit::ModelProcessConnection::ModelProcessConnection):
(WebKit::ModelProcessConnection::~ModelProcessConnection):
(WebKit::ModelProcessConnection::auditToken):
(WebKit::ModelProcessConnection::invalidate):
(WebKit::ModelProcessConnection::didClose):
(WebKit::ModelProcessConnection::didReceiveInvalidMessage):
(WebKit::ModelProcessConnection::dispatchMessage):
(WebKit::ModelProcessConnection::dispatchSyncMessage):
(WebKit::ModelProcessConnection::didInitialize):
(WebKit::ModelProcessConnection::waitForDidInitialize):
(WebKit::ModelProcessConnection::createVisibilityPropagationContextForPage):
(WebKit::ModelProcessConnection::destroyVisibilityPropagationContextForPage):
(WebKit::ModelProcessConnection::configureLoggingChannel):
* Source/WebKit/WebProcess/Model/ModelProcessConnection.h: Added.
(WebKit::ModelProcessConnection::connection):
(WebKit::ModelProcessConnection::messageReceiverMap):
(WebKit::ModelProcessConnection::Client::modelProcessConnectionDidClose):
(WebKit::ModelProcessConnection::addClient):
(WebKit::ModelProcessConnection::removeClient):
* Source/WebKit/WebProcess/Model/ModelProcessConnection.messages.in: Added.
* Source/WebKit/WebProcess/Model/ModelProcessConnectionInfo.h: Added.
(WebKit::ModelProcessConnectionInfo::encode const):
(WebKit::ModelProcessConnectionInfo::decode):
* Source/WebKit/WebProcess/Model/WebModelPlayer.h: Added.
(WebKit::WebModelPlayer::page):
(WebKit::WebModelPlayer::client):
* Source/WebKit/WebProcess/Model/WebModelPlayer.mm: Added.
(WebKit::WebModelPlayer::create):
(WebKit::WebModelPlayer::WebModelPlayer):
(WebKit::WebModelPlayer::~WebModelPlayer):
(WebKit::WebModelPlayer::load):
(WebKit::WebModelPlayer::sizeDidChange):
(WebKit::WebModelPlayer::layer):
(WebKit::WebModelPlayer::handleMouseDown):
(WebKit::WebModelPlayer::handleMouseMove):
(WebKit::WebModelPlayer::handleMouseUp):
(WebKit::WebModelPlayer::enterFullscreen):
(WebKit::WebModelPlayer::getCamera):
(WebKit::WebModelPlayer::setCamera):
(WebKit::WebModelPlayer::isPlayingAnimation):
(WebKit::WebModelPlayer::setAnimationIsPlaying):
(WebKit::WebModelPlayer::isLoopingAnimation):
(WebKit::WebModelPlayer::setIsLoopingAnimation):
(WebKit::WebModelPlayer::animationDuration):
(WebKit::WebModelPlayer::animationCurrentTime):
(WebKit::WebModelPlayer::setAnimationCurrentTime):
(WebKit::WebModelPlayer::hasAudio):
(WebKit::WebModelPlayer::isMuted):
(WebKit::WebModelPlayer::setIsMuted):
(WebKit::WebModelPlayer::accessibilityChildren):
* Source/WebKit/WebProcess/Model/WebModelPlayerProvider.cpp:
(WebKit::WebModelPlayerProvider::createModelPlayer):
Add a WebModelPlayer, which just launches a Model process and presents a green square.

* Source/WebKit/WebProcess/WebPage/WebPage.cpp:
(WebKit::WebPage::modelProcessConnectionDidBecomeAvailable):
(WebKit::WebPage::~WebPage):
(WebKit::WebPage::updatePreferences):
(WebKit::WebPage::configureLoggingChannel):
* Source/WebKit/WebProcess/WebPage/WebPage.h:
* Source/WebKit/WebProcess/WebProcess.cpp:
(WebKit::WebProcess::ensureModelProcessConnection):
(WebKit::WebProcess::modelProcessConnectionClosed):
* Source/WebKit/WebProcess/WebProcess.h:
(WebKit::WebProcess::existingModelProcessConnection):
</pre>

### CODE REVIEW

**Comment by hortont424** [comment]
This is just a rough first cut. I also generally followed existing patterns, even if they seemed questionable (especially re: naming oddities around "Process" vs. "Service" and similar).

File: Source/WebKit/ModelProcess/EntryPoint/Cocoa/XPCService/ModelService/Info-OSX.plist [hunk 1]
```diff
@@ -0,0 +1,49 @@
...
+		<key>JoinExistingSession</key>
+		<true/>
+		<key>RunLoopType</key>
+		<string>NSRunLoop</string>
+                <key>_ProcessType</key>
+                <string>App</string>
+		<key>_MultipleInstances</key>
+		<true/>
+	</dict>
+    <key>NSSupportsAutomaticGraphicsSwitching</key>
```
Comment (grorg) [inline Source/WebKit/ModelProcess/EntryPoint/Cocoa/XPCService/ModelService/Info-OSX.plist]: Do we still need this?

File: Source/WebKit/ModelProcess/EntryPoint/Cocoa/XPCService/ModelService/Info-OSX.plist [hunk 2]
```diff
@@ -0,0 +1,49 @@
...
+		<key>ServiceType</key>
+		<string>Application</string>
+		<key>JoinExistingSession</key>
+		<true/>
+		<key>RunLoopType</key>
+		<string>NSRunLoop</string>
+                <key>_ProcessType</key>
+                <string>App</string>
+		<key>_MultipleInstances</key>
+		<true/>
```
Comment (grorg) [inline Source/WebKit/ModelProcess/EntryPoint/Cocoa/XPCService/ModelService/Info-OSX.plist:44]: Does this mean multiple instances per parent process, or over the whole system?

File: Source/WebKit/Shared/Cocoa/DefaultWebBrowserChecks.mm [hunk 3]
```diff
@@ -71,7 +71,8 @@ static bool isInWebKitChildProcess()
         NSString *bundleIdentifier = [[NSBundle mainBundle] bundleIdentifier];
         isInSubProcess = [bundleIdentifier hasPrefix:@"com.apple.WebKit.WebContent"]
             || [bundleIdentifier hasPrefix:@"com.apple.WebKit.Networking"]
-            || [bundleIdentifier hasPrefix:@"com.apple.WebKit.GPU"];
+            || [bundleIdentifier hasPrefix:@"com.apple.WebKit.GPU"]
+            || [bundleIdentifier hasPrefix:@"com.apple.WebKit.Model"];
```
Comment (grorg) [inline Source/WebKit/Shared/Cocoa/DefaultWebBrowserChecks.mm]: We don't want to put the `ENABLE` guards around these things?

**Review by grorg — COMMENTED** [review]
I am not qualified!!! But since it's all boilerplate it looks fine to my eyes.

File: Source/WebKit/ModelProcess/EntryPoint/Cocoa/XPCService/ModelService/Info-OSX.plist (hunk 2 above)
Comment (hortont424) [reply Source/WebKit/ModelProcess/EntryPoint/Cocoa/XPCService/ModelService/Info-OSX.plist:44]: I'm not sure! Just stolen from the GPUP (which does the latter and not the former). It kind of looks like we don't actually use the SPI that this enables, anymore...

File: Source/WebKit/Shared/Cocoa/DefaultWebBrowserChecks.mm (hunk 3 above)
Comment (hortont424) [reply Source/WebKit/Shared/Cocoa/DefaultWebBrowserChecks.mm]: Hmm, I followed places that used ENABLE(GPU_PROCESS), not sure why they (may well have been me...) didn't here, we certainly can!

File: Source/WebKit/WebProcess/WebProcess.cpp [hunk 4]
```diff
@@ -1349,6 +1355,36 @@ AudioMediaStreamTrackRendererInternalUnitManager& WebProcess::audioMediaStreamTr
...
+
+void WebProcess::modelProcessConnectionClosed(ModelProcessConnection& connection)
+{
+    ASSERT(m_modelProcessConnection);
+    ASSERT_UNUSED(connection, m_modelProcessConnection == &connection);
+
+    m_modelProcessConnection = nullptr;
+}
+
+#endif // ENABLE(GPU_PROCESS)
```
Comment (hortont424) [inline Source/WebKit/WebProcess/WebProcess.cpp]: Missed a GPU->MODEL here

File: Source/WebKit/ModelProcess/ModelConnectionToWebProcess.cpp [hunk 5]
```diff
@@ -0,0 +1,204 @@
...
+#include "ModelProcessConnectionParameters.h"
+#include "ModelProcessMessages.h"
+#include "ModelProcessProxyMessages.h"
+#include "WebCoreArgumentCoders.h"
+#include "WebErrors.h"
+#include "WebProcessMessages.h"
+#include <WebCore/LogInitialization.h>
+
+#if PLATFORM(COCOA)
+#include <WebCore/SystemBattery.h>
```
Comment (hortont424) [inline Source/WebKit/ModelProcess/ModelConnectionToWebProcess.cpp]: Should not need this

File: Source/WebKit/WebKit.xcodeproj/project.pbxproj [hunk 6]
```diff
@@ -15897,6 +16103,72 @@
...
+			);
+			inputFileListPaths = (
+			);
+			inputPaths = (
+			);
+			name = "Create Symlink to Alt Root Path";
+			outputFileListPaths = (
+			);
+			outputPaths = (
+			);
```
Comment (aproskuryakov) [inline Source/WebKit/WebKit.xcodeproj/project.pbxproj:16149]: Why no inputs and outputs?

Comment (emw-apple) [reply Source/WebKit/WebKit.xcodeproj/project.pbxproj:16149]: I believe an output path of `$(OUTPUT_ALTERNATE_ROOT_PATH)`  is what you want here, as the script will run if and only if:

- OUTPUT_ALTERNATE_ROOT_PATH does not exst
- build settings (and therefore environment variables which may influence the symlink's destination) have changed

Unfortunately there is no consistent pattern across the project :(

File: Source/WebKit/WebKit.xcodeproj/project.pbxproj [hunk 7]
```diff
@@ -15897,6 +16103,72 @@
...
+			inputFileListPaths = (
+			);
+			inputPaths = (
+			);
+			name = "Copy XPC Service to Secondary Path";
+			outputFileListPaths = (
+			);
+			outputPaths = (
+			);
```
Comment (emw-apple) [inline Source/WebKit/WebKit.xcodeproj/project.pbxproj:16167]: Ditto above—we should have dependencies here—but IMO this is a problem with how folks use the script, not with this patch.

This has come up before in https://bugs.webkit.org/show_bug.cgi?id=237394. I finally filed https://bugs.webkit.org/show_bug.cgi?id=243913 to track the necessary script refactoring.

File: Source/WebKit/ModelProcess/EntryPoint/Cocoa/XPCService/ModelService/Info-iOS.plist [hunk 8]
```diff
@@ -0,0 +1,47 @@
...
+	<string>????</string>
+	<key>CFBundleVersion</key>
+	<string>${BUNDLE_VERSION}</string>
+	<key>LSFileQuarantineEnabled</key>
+	<true/>
+	<key>NSPrincipalClass</key>
+	<string>NSApplication</string>
+	<key>LSUIElement</key>
+	<true/>
+    <key>_STAttributionDisplayName</key>
```
Comment (weinig) [inline Source/WebKit/ModelProcess/EntryPoint/Cocoa/XPCService/ModelService/Info-iOS.plist:31]: Some funky spacing here.

File: Source/WebKit/ModelProcess/ModelProcess.cpp [hunk 9]
```diff
@@ -0,0 +1,263 @@
...
+constexpr Seconds minimumLifetimeBeforeIdleExit { 5_s };
+
+ModelProcess::ModelProcess(AuxiliaryProcessInitializationParameters&& parameters)
+    : m_idleExitTimer(*this, &ModelProcess::tryExitIfUnused)
+{
+    initialize(WTFMove(parameters));
+    RELEASE_LOG(Process, "%p - ModelProcess::ModelProcess:", this);
+}
+
+ModelProcess::~ModelProcess()
```
Comment (weinig) [inline Source/WebKit/ModelProcess/ModelProcess.cpp:70]: ModelProcess::~ModelProcess() = default;

File: Source/WebKit/WebProcess/Model/WebModelPlayer.h [hunk 10]
```diff
@@ -0,0 +1,83 @@
...
+#import "WebPageProxyMessages.h"
+#import <WebCore/ModelPlayer.h>
+#import <WebCore/ModelPlayerClient.h>
+#import <wtf/Compiler.h>
+
+OBJC_CLASS CALayer;
+
+namespace WebKit {
+
+class WebModelPlayer : public WebCore::ModelPlayer, public CanMakeWeakPtr<WebModelPlayer> {
```
Comment (weinig) [inline Source/WebKit/WebProcess/Model/WebModelPlayer.h:41]: Is this the ModelPlayer that will proxy to the ModelProcess? If so, maybe a better name would be something like ModelProcessModelPlayer?

File: Source/WebKit/WebProcess/Model/WebModelPlayer.h [hunk 11]
```diff
@@ -0,0 +1,83 @@
...
+#if ENABLE(MODEL_PROCESS)
+
+#import "ModelIdentifier.h"
+#import "WebPage.h"
+#import "WebPageProxyMessages.h"
+#import <WebCore/ModelPlayer.h>
+#import <WebCore/ModelPlayerClient.h>
+#import <wtf/Compiler.h>
+
+OBJC_CLASS CALayer;
```
Comment (weinig) [inline Source/WebKit/WebProcess/Model/WebModelPlayer.h:37]: Seems unnecessary to make this class/file require CoreAnimation as it stands right now.

**Comment by weinig** [comment]
I have a few high-level questions:

1) When you say "loading/parsing <model>s", is that intended to also include loading subresources? Or will that be something that is delegated to the WebProcess and then NetworkProcess still?
2) How much of this is duplicated GPU Process logic? If not immediately, do you see a path to reduce the duplication between them (base class, template base class, factoring out proxy logic, etc?) Or do you imagine they will diverge enough to not be worth it?
3) Can you expand a bit on the future lifetime expectations of the process? Is the intent to eventually have a one-to-one correspondence with model objects, or will a page with multiple models utilize a shared instance of the process?

On the last point, regardless of what approach to lifetime is taken, I think it would be good to think about how remove as much global state as possible (other than connection multiplexing) and consider not relying on the main run loop if possible, as it is much easier to make these kind of designs up front. It would also be good to think about how relevant WebPreferences might be passed to the process instances earlier rather than later, though that is also dependent on how you approach lifetime.

File: Source/WebKit/WebKit.xcodeproj/project.pbxproj (hunk 6 above)
Comment (hortont424) [reply Source/WebKit/WebKit.xcodeproj/project.pbxproj:16149]: >Why no inputs and outputs?

Cloned from GPUP, which also does not have them!

File: Source/WebKit/ModelProcess/EntryPoint/Cocoa/XPCService/ModelService/Info-iOS.plist (hunk 8 above)
Comment (hortont424) [reply Source/WebKit/ModelProcess/EntryPoint/Cocoa/XPCService/ModelService/Info-iOS.plist:31]: I don't see it, what am I missing?

File: Source/WebKit/WebProcess/Model/WebModelPlayer.h (hunk 11 above)
Comment (hortont424) [reply Source/WebKit/WebProcess/Model/WebModelPlayer.h:37]: Fair!

File: Source/WebKit/WebProcess/Model/WebModelPlayer.h (hunk 10 above)
Comment (hortont424) [reply Source/WebKit/WebProcess/Model/WebModelPlayer.h:41]: Yes. That seems like a better name, agreed.

**Comment by hortont424** [comment]
> I have a few high-level questions:
> 
> 1. When you say "loading/parsing s", is that intended to also include loading subresources? Or will that be something that is delegated to the WebProcess and then NetworkProcess still?

Actual resource loading will continue to work as usual, via the networking process. "Parsing" is really the right way to think of it.

> 2. How much of this is duplicated GPU Process logic? If not immediately, do you see a path to reduce the duplication between them (base class, template base class, factoring out proxy logic, etc?) Or do you imagine they will diverge enough to not be worth it?

99.9% of it is duplicated GPUP logic, and 99.9% of _that_ was duplicated Network process logic. I do think we should find ways to reduce duplication. So far it comes up so infrequently that not much thought has been put into it (besides what is already in e.g. AuxiliaryProcess). But this is a good opportunity to at least consider ways to improve.

> 3. Can you expand a bit on the future lifetime expectations of the process? Is the intent to eventually have a one-to-one correspondence with model objects, or will a page with multiple models utilize a shared instance of the process?

The intent (and current implementation) is one-model-process-per-web-process, though I could probably be talked up to one-model-process-per-WebPage (or -per-model-element). But **not** talked down to a GPUP-style singleton.

> On the last point, regardless of what approach to lifetime is taken, I think it would be good to think about how remove as much global state as possible (other than connection multiplexing) and consider not relying on the main run loop if possible, as it is much easier to make these kind of designs up front. 

Agreed!

> It would also be good to think about how relevant WebPreferences might be passed to the process instances earlier rather than later, though that is also dependent on how you approach lifetime.

Certainly all of the problems with WebPreferences go away in a _one-per-WebPage_ world, so maybe that's a better place to start? Or we just choose to send prefs that we care about (or the whole WebPreferences blob?) in the model-hosting-session-creation-message (does not exist yet). Either way seems fine.

**Comment by eddydas** [comment]
Taking this over from Tim.
Started a fresh PR from latest main: https://github.com/WebKit/WebKit/pull/23146

**Pull request closed without merging**